[DRAFT] OpenTelemetry investigation

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
TBD

Fixes: https://github.com/ClickHouse/ClickHouse/issues/52584

azat committed:
Add assertion to DistributedAsyncInsertHeader::createTracingContextHolder()

This should allow to catch the problem.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add test logging for opentelemetry

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: add SYSTEM STOP DISTRIBUTED SENDS like on the failed CI

Stopping distributed sends like increases the probability of catching
this error, since the query will be run only from the FLUSH context not
in some background thread.

CI: https://s3.amazonaws.com/clickhouse-test-reports/52567/0d8efed2871f399279df6a558da4b05cde1be69a/stateless_tests_flaky_check__asan_.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #52586 from azat/ClickHouse:opentelemetry-leftovers-fix -->
This is an automated comment for commit 1836d7536a73e4c670a7d1689ab7228993552ef4 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52586/1836d7536a73e4c670a7d1689ab7228993552ef4/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

azat committed:
FIXME

    (gdb) bt
    0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
    1  0x00007ffff7e2f2d3 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
    2  0x00007ffff7ddfa08 in __GI_raise (sig=sig@entry=6) at raise.c:26
    3  0x00007ffff7dc8538 in __GI_abort () at abort.c:79
    4  0x000000001b707faa in Poco::SignalHandler::handleSignal (sig=11) at SignalHandler.cpp:94
    5  <signal handler called>
    6  0xfffffffffffffff8 in ?? ()
    7  0x000000001246972a in DB::OwnFormattingChannel::logExtended (this=0x7ffff70a2540, msg=...) at OwnFormattingChannel.cpp:15
    8  0x000000001246e439 in DB::OwnSplitChannel::logSplit (this=this@entry=0x7ffff718ece0, msg=...) at OwnSplitChannel.cpp:88
    9  0x000000001246e2af in DB::OwnSplitChannel::tryLogSplit (this=this@entry=0x7ffff718ece0, msg=...) at OwnSplitChannel.cpp:54
    10 0x000000001246e1c5 in DB::OwnSplitChannel::log (this=0x7ffff718ece0, msg=...) at OwnSplitChannel.cpp:43
    11 0x00000000122dc91c in DB::OpenTelemetry::TracingContextHolder::~TracingContextHolder (this=0x7ffff5976e38) at OpenTelemetryTraceContext.cpp:379
    12 0x00000000122ceb42 in ThreadPoolImpl<std::__1::thread>::worker (this=0x7ffff7020300, thread_it=...) at ThreadPool.cpp:460
    (gdb) info threads
      Id   Target Id                        Frame
    * 1    Thread 0x7ffff597f6c0 (LWP 3504) __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
      2    Thread 0x7ffff7cb4700 (LWP 3329) warning: Section `.reg-xstate/3329' in core file too small.
    0x00007ffff7e29f0e in __futex_abstimed_wait_common64 (private=128, cancel=true, abstime=0x0, op=265, expected=3504, futex_word=0x7ffff597f990) at futex-internal.c:57
    (gdb) t 2
    [Switching to thread 2 (Thread 0x7ffff7cb4700 (LWP 3329))]
    0  0x00007ffff7e29f0e in __futex_abstimed_wait_common64 (private=128, cancel=true, abstime=0x0, op=265, expected=3504, futex_word=0x7ffff597f990) at futex-internal.c:57
    57          return INTERNAL_SYSCALL_CANCEL (futex_time64, futex_word, op, expected,
    gdb) bt
    0  0x00007ffff7e29f0e in __futex_abstimed_wait_common64 (private=128, cancel=true, abstime=0x0, op=265, expected=3504, futex_word=0x7ffff597f990) at futex-internal.c:57
    1  __futex_abstimed_wait_common (futex_word=futex_word@entry=0x7ffff597f990, expected=3504, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=128, cancel=cancel@entry=true) at futex-internal.c:87
    2  0x00007ffff7e29f8f in __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x7ffff597f990, expected=<optimized out>, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=128) at futex-internal.c:139
    3  0x00007ffff7e2f023 in __pthread_clockjoin_ex (threadid=140737313765056, thread_return=0x0, clockid=0, abstime=0x0, block=<optimized out>) at pthread_join_common.c:102
    4  0x000000001dc44a93 in std::__1::__libcpp_thread_join[abi:v15000](unsigned long*) (__t=0x7ffff707b830) at __threading_support:396
    5  std::__1::thread::join (this=0x7ffff707b830) at thread.cpp:49
    6  0x00000000122cd9e9 in ThreadPoolImpl<std::__1::thread>::finalize (this=this@entry=0x7ffff7020300) at ThreadPool.cpp:297
    7  0x00000000122d5850 in ThreadPoolImpl<std::__1::thread>::~ThreadPoolImpl (this=0x7ffff597f990, this@entry=0x7ffff7020300) at ThreadPool.cpp:278
    8  0x00000000122d1c78 in std::__1::default_delete<GlobalThreadPool>::operator()[abi:v15000](GlobalThreadPool*) const (this=<optimized out>, __ptr=0x7ffff7020300) at unique_ptr.h:48
    9  std::__1::unique_ptr<GlobalThreadPool, std::__1::default_delete<GlobalThreadPool> >::reset[abi:v15000](GlobalThreadPool*) (this=<optimized out>, __p=0x0) at unique_ptr.h:305
    10 std::__1::unique_ptr<GlobalThreadPool, std::__1::default_delete<GlobalThreadPool> >::~unique_ptr[abi:v15000]() (this=<optimized out>) at unique_ptr.h:259
    11 0x00007ffff7de2066 in __run_exit_handlers (status=0, listp=0x7ffff7f81760 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:111
    12 0x00007ffff7de21b0 in __GI_exit (status=<optimized out>) at exit.c:141
    13 0x00007ffff7dc9857 in __libc_start_call_main (main=main@entry=0xd38c780 <main(int, char**)>, argc=argc@entry=7, argv=argv@entry=0x7fffffffc7e8) at libc_start_call_main.h:74
    14 0x00007ffff7dc990a in __libc_start_main_impl (main=0xd38c780 <main(int, char**)>, argc=7, argv=0x7fffffffc7e8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffc7d8) at libc-start.c:360
    15 0x000000000d38c52e in _start ()
    (gdb) quit

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Move logging in dtor to avoid SIGSEGV

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Replace assert with LOGICAL_ERROR

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Adjust messages

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Note

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Logging

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add missing LOGICAL_ERRROR extern

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
